Improved recipe initialization UX (#9158)
PrefectHQ/prefect: docs/concepts/projects.md

### The Build Section

> The build section of `prefect.yaml` is where any necessary side effects for running your deployments are built - the most common type of side effect produced here is a Docker image.  If we initialize with the docker recipe we are met with a series of prompts to provide required information such as image name and tag: 

Reviewer: ```suggestion
The build section of `prefect.yaml` is where any necessary side effects for running your deployments are built - the most common type of side effect produced here is a Docker image.  If you initialize with the docker recipe, you will be prompted to provide required information, such as image name and tag: 
```

<div class="terminal">
```bash
$ prefect project init --recipe docker
>> image_name: < insert image name here >
>> tag: < insert image tag here >
```
</div>

!!! tip "Use `--field` to avoid the interactive experience"
    We recommend that you only initialize a recipe when you begin a project, and afterwards store your configuration files within version control; however, sometimes you may need to initialize programmatically and avoid the interactive prompts.  To do so, provide all required fields for your recipe using the `--field` flag:
    <div class="terminal">
    ```bash
    $ prefect project init --recipe docker \
        --field image_name=my-repo/my-image \
        --field tag=my-tag
    ```
    </div>
    
```yaml
build:
- prefect_docker.projects.steps.build_docker_image:
    requires: prefect-docker>=0.2.0
    image_name: my-repo/my-image
    tag: my-tag
    dockerfile: auto
```

> Once we confirm that these fields are set to their desired values, this step will automatically build a Docker image with the provided name and tag and push it to the repository referenced by the image name.  [As the documentation notes](https://prefecthq.github.io/prefect-docker/projects/steps/#prefect_docker.projects.steps.BuildDockerImageResult), this step produces a few fields that can optionally be used in future steps or within `deployment.yaml` as template values.  It is best practice to use `{{ image_name }}` within `deployment.yaml` (specificially the work pool's job variables section) so that you don't risk having your build step and deployment specification get out of sync with hardcoded values.  For a worked example, [check out the project tutorial](/tutorials/projects/#dockerized-deployment).

Reviewer: ```suggestion
Once you've confirmed that these fields are set to their desired values, this step will automatically build a Docker image with the provided name and tag and push it to the repository referenced by the image name.  [As the documentation notes](https://prefecthq.github.io/prefect-docker/projects/steps/#prefect_docker.projects.steps.BuildDockerImageResult), this step produces a few fields that can optionally be used in future steps or within `deployment.yaml` as template values.  It is best practice to use `{{ image_name }}` within `deployment.yaml` (specificially the work pool's job variables section) so that you don't risk having your build step and deployment specification get out of sync with hardcoded values.  For a worked example, [check out the project tutorial](/tutorials/projects/#dockerized-deployment).
```

The pull section is the most important section within the `prefect.yaml` file as it contains instructions for preparing this project for a deployment run.  These instructions will be executed each time a deployment created wthin this project is run via a worker.

There are three main types of steps that typically show up in a `pull` section:

- `set_working_directory`: this step simply sets the working directory for the process prior to importing your flow
- `git_clone_project`: this step clones the provided repository on the provided branch
- `pull_project_from_{cloud}`: this step pulls the project directory from a Cloud storage location (e.g., S3)

!!! tip "Use block and variable references"

>     All [block and variable references](#templating-options) within your pull step will remain unresolved until runtime and will be pulled each time your deployment is run. This allows you to avoid storing sensitive information in an insecure way; it also allows you to manage certain types of configuration from the API and UI without having to rebuild your deployment every time.

Reviewer: ```suggestion
    All [block and variable references](#templating-options) within your pull step will remain unresolved until runtime and will be pulled each time your deployment is run. This allows you to avoid storing sensitive information insecurely; it also allows you to manage certain types of configuration from the API and UI without having to rebuild your deployment every time.
```
